src/OpenTelemetry.Instrumentation.AspNet.Win/Implementation/HttpInListener.cs

```diff
@@ -165,6 +167,13 @@ public override void OnStopActivity(Activity activity, object payload)
                     activity.AddTag(SpanAttributeConstants.HttpRouteKey, template);
                 }
             }
+
+            if (activity.OperationName.Equals(ActivityNameByHttpInListener))
```

Comment:
@cijothomas This is confusing me. We're in OnStopActivity but we need to call activity.Stop()? Isn't that what triggers this event?

I feel like this might be making some assumptions about how library authors will call us that might not necessarily be true. Forgive me if I'm totally wrong and confused!

Flow A (bugged but could work?)...
* Library creates and starts Activity.
* We create a new one due to the new logic. Activity.Current is now ours.
* Library calls Activity.Current.Stop(). They think they are stopping their instance, but it is really ours.
* Activity.Stop calls Activity.SetCurrent(Parent). This sets Activity.Current back to the original Activity.
* This event fires, using Activity.Current (which is now the original).
   * activity parameter is now the parent activity, still running.
   * Our check `if (activity.OperationName.Equals(ActivityNameByHttpInListener))` is false? We don't stop anything, but we should. Easy fix.

Flow B (not working)...
* Library creates and starts Activity.
* We create a new one due to the new logic. Activity.Current is now ours.
* Library calls activity.Stop on their instance when they are done. The original parent Activity is now stopped. Ours is still running.
* Activity.Stop calls Activity.SetCurrent(Parent). Activity.Current is now unknown parent before this ever started. Could be nothing, could be something.
* This event fires, using Activity.Current.
   * activity parameter is now the parent of the original activity.
   * Our Activity is still running but we have no way to stop it?

Reply:
Flow A - Its not correct as we call activity.stop before library calls Actiivty.Current.Stop()

Flow B - Again, we call activity.stop before library call their .Stop.


The following is the correct ordering per my understanding:

Library does    var myActivity = Activity("library").Start
Library does   DiagnosticSource.Write  --> OT gets call back Activity.Current = Activity("library") - But OT does Activity("OT").Start. 
Now Activity.Current = Activity("OT")

..do job..

Library does   DiagnosticSource.Write  -- OT gets call back Activity.Current = Activity("OT")      - OT does Activity("OT").Stop
Now Activity.Current becomes null.

If library does   Activity.Current?.Stop() - the myActivity never gets stopped. (asp.net does this)
If library does myActivity.Stop() - we are good. (asp.net core does this)

Looks like we'll have an issue in Asp.Net. We can discuss this, but please confirm if my above understanding is correct.

Reply:
Let me play with it a little bit tomorrow and I'll get back to you. One idea I just had... in the OnStartActivity logic, what if we did this...

```csharp
            if (!(this.options.TextFormat is TraceContextFormatActivity))
            {
                ...
                Activity newOne = new Activity(ActivityNameByHttpInListener);
                ...
                activity.CustomProperties("ot.childActivity", newOne); // Add child to parent.
                activity = newOne;
            }
```

We add a reference to our created child span on the parent so if we get the parent in OnStopActivity (instead of the child) we can still clean everything up?

Reply:
We'll need to do something only for Asp.Net case right? 
I think what we need is then to store the libraryActivity into newOne.SetCustomProperty. So OT adapter can stop both ours, and the libraryActivity. Because we anyway have to start newOne, as that'll ensure it becomes current. (else child spans like httpclient/sqlclient etc will not have correct id)


Before that - do you agree that this issue is only for Asp.Net (or any libraries which does Activity.Current.Stop? )

Reply:
> do you agree that this issue is only for Asp.Net

Not sure yet if there is even an issue. I want to make a test library that I can mess with how Activity is stopped.

> or any libraries which does Activity.Current.Stop

That's what I'm more worried about. The Asp.Net stuff we can just fix what it does. But ideally we have a robust solution where library author doesn't need to worry about how they stop their activities.

Reply:
@cijothomas OK I got a second to actually open the solution and look at the code. You are correct it's just an issue with the Asp.Net HttpInListener. I thought for some reason this was universal logic but it only applies to specific DiagnosticSources so we don't need to worry about libraries other than what it's targeting! Sorry about that.

> I think what we need is then to store the libraryActivity into newOne.SetCustomProperty. So OT adapter can stop both ours, and the libraryActivity.

I got your branch up and running and then set a custom `ITextFormatActivity` via the options in Global.asax.cs in the OpenTelemetry.Exporter.AspNet.Win sample project. When I send a request through I hit the new logic (`new Activity(ActivityNameByHttpInListener)`. When `OnStopActivity` fires, activity.OperationName is `Microsoft.AspNet.HttpReqIn`. So I think we need to go the route of adding the "newOne" to the "libraryActivity"? But either way, good news we can make this work.

The "CustomFormat" unit test works for both, so I think something is off there? It should fail as-is for Asp.Net right?

Reply:
By the way, I did the same test in OpenTelemetry.Exporter.Web, adding a custom `ITextFormatActivity` in Startup.cs. It works like a champ. When I send a request through I hit the new logic (new Activity(ActivityNameByHttpInListener). When OnStopActivity fires, activity.OperationName is ActivityCreatedByHttpInListener and the "newOne" is stopped.

Reply:
I think Asp.Net is doing something special. It has some special logic to restore current activity. This must be why you can see `Microsoft.AspNet.HttpReqIn` in `OnStopActivity`.

If its restoring Current to the libraryActivity, then we need to stop the one created by HttpInListener inside libraryActivity.

If its behaviour is like what we mock in unit test (shown below), then we need to store the libraryActivity inside the one created by HttpInListener. 

```
activity.Start();
                this.fakeAspNetDiagnosticSource.Write(
                    "Start",
                    null);

                this.fakeAspNetDiagnosticSource.Write(
                    "Stop",
                    null);
                activity.Stop();
```

Need to investigate this a bit more.

Reply:
https://github.com/aspnet/Microsoft.AspNet.TelemetryCorrelation/blob/master/src/Microsoft.AspNet.TelemetryCorrelation/ActivityHelper.cs#L101 - This explains the behavior we are seeing.

Reply:
Made a quick and hacky fix - I'm not very sure if this is the right approach. Will investigate more.

Reply:
Essentially have both activity store each other - so if Current is restored or not, we can still access the other one given one! And always stop the one we started, and restore Current back to original. Sounds a lot of hacks...